Added font fullname to face fields (to include font style).

pkulchenko opened the pull request:
Simon, per the discussion in #95, this adds `fullname` fields to the `face` result for the font substitution, which allows to properly load the font (to use with alternative viewers) as `family` only returns the name, but doesn't return the style.

From the documentation:

```
 fullname       FC_FULLNAME            String  Font face full name where
                                               different from family and
                                               family + style
```

alerque commented:
@pkulchenko Out of curiosity were you able to replicate the error Travis CI ran into locally?

	### Compiling examples/showoff.sil
	This is SILE 0.9.3-unreleased
	<examples/showoff.sil>
	Error detected:
	core/harfbuzz-shaper.lua:28: attempt to index local 'face' (a nil value)
	The command "perl tests/attack.pl" exited with 1.

Just wondering if this is a problem with the tests or with something showing up in edge cases that your changes broke.

pkulchenko commented:
@alerque, I've seen this error before, but I don't think it's related to the changes I made. I have two different configurations of fontconfig and one of them doesn't have any config; this is what I get when I run the example with that configuration:

```
<examples\showoff.sil>Fontconfig error: Cannot load default config file

Error detected:
core/harfbuzz-shaper.lua:28: attempt to index local 'face' (a nil value)
```

However, when I run with the complete fontconfig setup, I don't have any issues and everything works as expected. I also tried to complete roll back the changes and I get the same error, so it doesn't seem to be related to what I'm doing.

It's strange that Travis reports this error, so there is still a chance that I'm doing something wrong. Have you tried to compile/run this branch locally?

pkulchenko commented:
@simoncozens, can you review the changes when you get a chance? They are quite simple and I'm not sure why the build would be failing as it compiles and runs for me locally without any issues (and I can get the information I need based on the changes).

simoncozens commented:
I'm getting the error and I think I see why. If the match totally fails then fullname is empty and so `return 0` happens without returning a table to the Lua stack. Here's more debugging output:

```
[fonts]	Resolved font family Gill Sans -> /Library/Fonts/GillSans.ttc
Gill Sans;10;200;normal;normal;;LTR
[fonts]	Resolved font family Gill Sans -> nil

! Could not find requested font {direction=LTR,features=,font=Gill Sans,language=en,script=,size=10,style=normal,variant=normal,weight=200} or any suitable substitutes at examples/showoff.sil
```

simoncozens commented:
OK, the match *is* working and finding Gill Sans Light, but the call to get `FC_FULLNAME` is failing. I still don't understand why that call is failing but rewriting it as

```C
  if (FcPatternGetString (matched, FC_FULLNAME, 0, &fullname) != FcResultMatch) {
    lua_pushstring(L, "fullname");
    lua_pushstring(L, (char*)(fullname));
    lua_settable(L, -3);
  }
```

does the right thing.

pkulchenko commented:
Ok; that explains the error. Strange though as it seems like it may match `familyname`, but not `fullname`? I was about to suggest the change, but you seem already implemented it ;). Do you want me to update the pull request?

simoncozens commented:
Yes please!

simoncozens commented:
Also: I strongly suspect the fact that Gill Sans is a TTC file is the problem here. But you have `index` so you can choose the right font within the collection.

pkulchenko commented:
`index` won't help in my case as wxwidgets doesn't give me access to this information, but I can always do `face.fullname or face.family`.

Pushed the changes...

pkulchenko commented:
It passes both Travis and my own tests, so we should be good with the change.

pkulchenko commented:
Simon, I noticed one interesting issue. `fullname` reports one font that is not available on my system:

```
! Font 'Frutiger LT Std' not available, falling back to 'Agency FB'
[wxlua-output]	Ignored substituted font 'Agency FB Negreta' as it's missing
```

Note that the family name is `Agency FB`, but the full name is reported as `Agency FB Negreta`. I don't have such a font installed (and it doesn't load when I try loading it using the full name).

I fixed this in my output module by checking against the list of known fonts (and added a warning message as shown above), but I thought I should mention this. I checked the C code and it looks correct, but maybe `Negreta` is some leftover from a previous call in the buffer? I don't know where else it would come from...

simoncozens commented:
No, you're fine. It's this: https://bugs.freedesktop.org/show_bug.cgi?id=27765

pkulchenko commented:
ok; so it should go away with the next fontconfig update (or I just need to compile it from sources).
